File: frontend/src/app/model/application/type/application-kind.ts
import {KindsWithSpecifiers, SpecifierEntry, toKindsWithSpecifiers as specifiersToKindsWithSpecifiers} from './application-specifier';

export enum ApplicationKind {
  PROMOTION = 'PROMOTION',
  OUTDOOREVENT = 'OUTDOOREVENT',
  BIG_EVENT = 'BIG_EVENT', // Suuret tapahtumat
  BRIDGE_BANNER = 'BRIDGE_BANNER', // Banderollit silloissa
  BENJI = 'BENJI', // Benji-hyppylaite
  PROMOTION_OR_SALES = 'PROMOTION_OR_SALES', // Esittely- tai myyntitila liikkeen edustalla
  URBAN_FARMING = 'URBAN_FARMING', // Kaupunkiviljelypaikka
  KESKUSKATU_SALES = 'KESKUSKATU_SALES', // Keskuskadun myyntipaikka
  SUMMER_THEATER = 'SUMMER_THEATER', // Kesäteatterit
  DOG_TRAINING_FIELD = 'DOG_TRAINING_FIELD', // Koirakoulutuskentät
  DOG_TRAINING_EVENT = 'DOG_TRAINING_EVENT', // Koirakoulutustapahtuma
  SMALL_ART_AND_CULTURE = 'SMALL_ART_AND_CULTURE', // Pienimuotoinen taide- ja kulttuuritoiminta
  SEASON_SALE = 'SEASON_SALE', // Sesonkimyynti
  CIRCUS = 'CIRCUS', // Sirkus/tivolivierailu
  ART = 'ART', // Taideteos
  STORAGE_AREA = 'STORAGE_AREA', // Varastoalue
  SUMMER_TERRACE = 'SUMMER_TERRACE', // Kesäterassi
  WINTER_TERRACE = 'WINTER_TERRACE', // Talviterassi
  PARKLET = 'PARKLET', // Parklet
  MOBILE_SALES = 'MOBILE_SALES', // Liikkuva myynti/myyntiautot ja -vaunut
  STREET_AND_GREEN = 'STREET_AND_GREEN', // Katu- ja vihertyöt
  WATER_AND_SEWAGE = 'WATER_AND_SEWAGE', // Vesi / viemäri
  ELECTRICITY = 'ELECTRICITY', // Sähkö
  DATA_TRANSFER = 'DATA_TRANSFER', // Tiedonsiirto
  HEATING_COOLING = 'HEATING_COOLING', // Lämmitys/viilennys
  CONSTRUCTION = 'CONSTRUCTION', // Rakennus
  YARD = 'YARD', // Piha
  GEOLOGICAL_SURVEY = 'GEOLOGICAL_SURVEY', // Pohjatutkimus
  PROPERTY_RENOVATION = 'PROPERTY_RENOVATION', // kiinteistöremontti
  CONTAINER_BARRACK = 'CONTAINER_BARRACK', // kontti/parakki
  PHOTO_SHOOTING = 'PHOTO_SHOOTING', // kuvaus
  SNOW_WORK = 'SNOW_WORK', // lumenpudotus
  RELOCATION = 'RELOCATION', // muutto
  LIFTING = 'LIFTING', // nostotyö
  NEW_BUILDING_CONSTRUCTION = 'NEW_BUILDING_CONSTRUCTION', // uudisrakennuksen työmaa-alue
  ROLL_OFF = 'ROLL_OFF', // vaihtolava
  CHRISTMAS_TREE_SALES_AREA = 'CHRISTMAS_TREE_SALES_AREA', // Joulukuusenmyyntipaikka
  CITY_CYCLING_AREA = 'CITY_CYCLING_AREA', // Kaupunkipyöräpaikka
  AGILE_KIOSK_AREA = 'AGILE_KIOSK_AREA', // Ketterien kioskien myyntipaikka
  STATEMENT = 'STATEMENT', // Lausunto
  SNOW_HEAP_AREA = 'SNOW_HEAP_AREA', // Lumenkasauspaikka
  SNOW_GATHER_AREA = 'SNOW_GATHER_AREA', // Lumenvastaanottopaikka
  OTHER_SUBVISION_OF_STATE_AREA = 'OTHER_SUBVISION_OF_STATE_AREA', // Muun hallintokunnan alue
  MILITARY_EXCERCISE = 'MILITARY_EXCERCISE', // Sotaharjoitus
  WINTER_PARKING = 'WINTER_PARKING', // Talvipysäköinti
  REPAVING = 'REPAVING', // Uudelleenpäällystykset
  ELECTION_ADD_STAND = 'ELECTION_ADD_STAND', // Vaalimainosteline
  PUBLIC_EVENT = 'PUBLIC_EVENT', // Yleisötilaisuus
  OTHER = 'OTHER'
}

export const terraceKinds: ApplicationKind[] = [
  ApplicationKind.SUMMER_TERRACE,
  ApplicationKind.WINTER_TERRACE,
  ApplicationKind.PARKLET
];

export const disabledKinds: ApplicationKind[] = [
  ApplicationKind.DOG_TRAINING_EVENT
];

export const commonKinds: KindsWithSpecifiers = {
  STREET_AND_GREEN: [
    'ASPHALT',
    'INDUCTION_LOOP',
    'COVER_STRUCTURE',
    'STREET_OR_PARK',
    'PAVEMENT',
    'TRAFFIC_LIGHT',
    'COMMERCIAL_DEVICE',
    'TRAFFIC_STOP',
    'BRIDGE',
    'OUTDOOR_LIGHTING'
  ],
  WATER_AND_SEWAGE: [
    'STORM_DRAIN',
    'WELL',
    'UNDERGROUND_DRAIN',
    'WATER_PIPE',
    'DRAIN'
  ],
  ELECTRICITY: [
    'DISTRIBUTION_CABINET',
    'ELECTRICITY_CABLE',
    'ELECTRICITY_WELL'
  ],
  DATA_TRANSFER: [
    'DISTRIBUTION_CABINET_OR_PILAR',
    'DATA_CABLE',
    'DATA_WELL'
  ],
  HEATING_COOLING: [
    'STREET_HEATING',
    'DISTRICT_HEATING',
    'DISTRICT_COOLING',
    'GEO_HEATING'
  ],
  CONSTRUCTION: [
    'GROUND_ROCK_ANCHOR',
    'UNDERGROUND_STRUCTURE',
    'UNDERGROUND_SPACE',
    'BASE_STRUCTURES',
    'DRILL_PILE',
    'CONSTRUCTION_EQUIPMENT',
    'CONSTRUCTION_PART',
    'GROUND_FROST_INSULATION',
    'SMOKE_HATCH_OR_PIPE',
    'STOP_OR_TRANSITION_SLAB',
    'SUPPORTING_WALL_OR_PILE'
  ],
  YARD: [
    'FENCE_OR_WALL',
    'DRIVEWAY',
    'STAIRS_RAMP',
    'SUPPORTING_WALL_OR_BANK'
  ],
  GEOLOGICAL_SURVEY: [
    'DRILLING',
    'TEST_HOLE',
    'GROUND_WATER_PIPE'
  ],
  OTHER: [
    'ABSORBING_SEWAGE_SYSTEM',
    'GAS_PIPE',
    'OTHER'
  ]
};

export function drawingAllowedForKind(kind: ApplicationKind): boolean {
  return ![ApplicationKind.BRIDGE_BANNER, ApplicationKind.DOG_TRAINING_EVENT, ApplicationKind.DOG_TRAINING_FIELD]
    .some(k => k === kind);
}

export function mergeKindsWithSpecifiers(kinds: ApplicationKind[], specifierEntries: SpecifierEntry[]): KindsWithSpecifiers {
  return {
    ...toKindsWithSpecifiers(kinds),
    ...specifiersToKindsWithSpecifiers(specifierEntries)
  };
}

function toKindsWithSpecifiers(kinds: ApplicationKind[]): KindsWithSpecifiers {
  return kinds.reduce((prev, cur) => ({
    ...prev,
    ...{[cur]: []}
  }), {});
}

File: backend/common-domain/src/main/java/fi/hel/allu/common/domain/types/ApplicationKind.java
package fi.hel.allu.common.domain.types;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Application kind specifies the subtype of application.
 */
public enum ApplicationKind {
  PROMOTION(ApplicationType.EVENT),
  OUTDOOREVENT(ApplicationType.EVENT),
  BIG_EVENT(ApplicationType.EVENT),
  BRIDGE_BANNER(ApplicationType.SHORT_TERM_RENTAL), // Banderollit silloissa
  BENJI(ApplicationType.SHORT_TERM_RENTAL), // Benji-hyppylaite
  PROMOTION_OR_SALES(ApplicationType.SHORT_TERM_RENTAL), // Esittely- tai myyntitila liikkeen edustalla
  URBAN_FARMING(ApplicationType.SHORT_TERM_RENTAL), // Kaupunkiviljelypaikka
  KESKUSKATU_SALES(ApplicationType.SHORT_TERM_RENTAL), // Keskuskadun myyntipaikka
  SUMMER_THEATER(ApplicationType.SHORT_TERM_RENTAL), // Kesäteatterit
  DOG_TRAINING_FIELD(ApplicationType.SHORT_TERM_RENTAL), // Koirakoulutuskentät
  DOG_TRAINING_EVENT(ApplicationType.SHORT_TERM_RENTAL), // Koirakoulutustapahtuma
  SMALL_ART_AND_CULTURE(ApplicationType.SHORT_TERM_RENTAL), // Pienimuotoinen taide- ja kulttuuritoiminta
  SEASON_SALE(ApplicationType.SHORT_TERM_RENTAL), // Sesonkimyynti
  CIRCUS(ApplicationType.SHORT_TERM_RENTAL), // Sirkus/tivolivierailu
  ART(ApplicationType.SHORT_TERM_RENTAL), // Taideteos
  STORAGE_AREA(ApplicationType.SHORT_TERM_RENTAL, ApplicationType.AREA_RENTAL), // Varastoalue
  SUMMER_TERRACE(ApplicationType.SHORT_TERM_RENTAL), // Kesäterassi
  WINTER_TERRACE(ApplicationType.SHORT_TERM_RENTAL), // Talviterassi
  PARKLET(ApplicationType.SHORT_TERM_RENTAL), // Parklet
  MOBILE_SALES(ApplicationType.SHORT_TERM_RENTAL), // Liikkuva myynti/myyntiautot ja -vaunut
  STREET_AND_GREEN(ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT,
      ApplicationType.PLACEMENT_CONTRACT), // Katu- ja vihertyöt
  WATER_AND_SEWAGE(ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT,
      ApplicationType.PLACEMENT_CONTRACT), // Vesi / viemäri
  ELECTRICITY(ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT,
      ApplicationType.PLACEMENT_CONTRACT), // Sähkö
  DATA_TRANSFER(ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT,
      ApplicationType.PLACEMENT_CONTRACT), // Tiedonsiirto
  HEATING_COOLING(ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT,
      ApplicationType.PLACEMENT_CONTRACT), // Lämmitys/viilennys
  CONSTRUCTION(ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT,
      ApplicationType.PLACEMENT_CONTRACT), // Rakennus
  YARD(ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT, ApplicationType.PLACEMENT_CONTRACT), // Piha
  GEOLOGICAL_SURVEY(ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT,
      ApplicationType.PLACEMENT_CONTRACT), // Pohjatutkimus
  PROPERTY_RENOVATION(ApplicationType.AREA_RENTAL),        // kiinteistöremontti
  CONTAINER_BARRACK(ApplicationType.AREA_RENTAL),          // kontti/parakki
  PHOTO_SHOOTING(ApplicationType.AREA_RENTAL),             // kuvaus
  SNOW_WORK(ApplicationType.AREA_RENTAL),                  // lumenpudotus
  RELOCATION(ApplicationType.AREA_RENTAL),                 // muutto
  LIFTING(ApplicationType.AREA_RENTAL),                    // nostotyö
  NEW_BUILDING_CONSTRUCTION(ApplicationType.AREA_RENTAL),  // uudisrakennuksen työmaa-alue
  ROLL_OFF(ApplicationType.AREA_RENTAL),                   // vaihtolava
  CHRISTMAS_TREE_SALES_AREA(ApplicationType.NOTE), // Joulukuusenmyyntipaikka
  CITY_CYCLING_AREA(ApplicationType.NOTE), // Kaupunkipyöräpaikka
  AGILE_KIOSK_AREA(ApplicationType.NOTE), // Ketterien kioskien myyntipaikka
  STATEMENT(ApplicationType.NOTE), // Lausunto
  SNOW_HEAP_AREA(ApplicationType.NOTE), // Lumenkasauspaikka
  SNOW_GATHER_AREA(ApplicationType.NOTE), // Lumenvastaanottopaikka
  OTHER_SUBVISION_OF_STATE_AREA(ApplicationType.NOTE), // Muun hallintokunnan alue
  MILITARY_EXCERCISE(ApplicationType.NOTE), // Sotaharjoitus
  WINTER_PARKING(ApplicationType.NOTE), // Talvipysäköinti
  REPAVING(ApplicationType.NOTE), // Uudelleenpäällystykset
  ELECTION_ADD_STAND(ApplicationType.NOTE), // Vaalimainosteline
  PUBLIC_EVENT(ApplicationType.TEMPORARY_TRAFFIC_ARRANGEMENTS, ApplicationType.AREA_RENTAL), // Yleisötilaisuus
  OTHER(ApplicationType.SHORT_TERM_RENTAL, ApplicationType.CABLE_REPORT, ApplicationType.EXCAVATION_ANNOUNCEMENT,
      ApplicationType.PLACEMENT_CONTRACT, ApplicationType.AREA_RENTAL,
      ApplicationType.NOTE, ApplicationType.TEMPORARY_TRAFFIC_ARRANGEMENTS); // Muu

  private final List<ApplicationType> types;

  /**
   * Get the allowed application type for this application kind
   */
  public List<ApplicationType> getTypes() {
    return types;
  }

  private ApplicationKind(ApplicationType... types) {
    this.types = Arrays.asList(types);
  }

  public static List<ApplicationKind> forApplicationType(ApplicationType type) {
    return Arrays.stream(ApplicationKind.values()).filter(k -> k.types.contains(type)).collect(Collectors.toList());
  }

  public boolean isTerrace() {
    return this == SUMMER_TERRACE || this == WINTER_TERRACE || this == PARKLET;
  }
}
